### PR TITLE
Create items, collaborations on folders, folder recursive delete

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by bcoe** [comment]
This looks great! I'll pull the code in over the next few days and write some integration tests.

Do you think you could add some unit tests into folder_spec.rb for this new functionality. I usually:
- do some manual testing in irb.
- capture some JSON responses and save them in the fixtures directory.
- write tests around the fixtures.

thanks for the contribution :+1:

**Comment by ghost** [comment]
Hi Benjamin,

  I normally would put some unit test around any code I submit, but we're
just working at an insane pace right now here at work.  As soon as we
deliver for our upcoming demo, I'll go back and do this.  Of note, we're
also working on adding some code to have the gem auto-refresh access tokens
or at least provide access to figure out if the token has expired already.

On Thu, May 23, 2013 at 2:05 PM, Benjamin E. Coe
notifications@github.comwrote:

> This looks great! I'll pull the code in over the next few days and write
> some integration tests.
> 
> Do you think you could add some unit tests into folder_spec.rb for this
> new functionality. I usually:
> - do some manual testing in irb.
> - capture some JSON responses and save them in the fixtures directory.
> - write tests around the fixtures.
> 
> thanks for the contribution [image: :+1:]
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/attachmentsme/ruby-box/pull/35#issuecomment-18361017
> .

## 

*

Roland Lopes
Agile Engineer
Tel:    416.879.5460
Email: rlopes@xtremelabs.com
Skype: roland-xtreme

*

**Comment by bcoe** [comment]
Cool, I'd work from your fork of the repo for the time being. If I have the time, I'll write some specs and pull this code in within the next few days.

File: lib/ruby-box/session.rb [hunk 1]
```diff
@@ -66,7 +67,23 @@ def request(uri, request, raw=false)
       if response.is_a? Net::HTTPNotFound
         raise RubyBox::ObjectNotFound
       end
+
+      # Got unauthorized (401) status, try to refresh the token
+      if response.code.to_i == 401
```
Comment (bcoe) [inline lib/ruby-box/session.rb:72]: If we're going to be backward compatible wit the old API  api_key based approach to interacting with the Box API, we should check to make sure that a @refresh_token actually exists, prior to attempting to refresh it.

Do you mind re-submitting your collaboration pull request on a branch named 'collaboration', would love to pull that stuff in separately.

**Comment by bcoe** [comment]
Thanks for the great contributions to ruby-box @xtreme-roland-lopes I've pulled in all of your changes with a few minor modifications:
- rather than adding a recursive delete method to folder, you can now pass in options to delete so: folder.delete(recursive: 'true') will provide the recursive delete functionality.
- I didn't like silently suppressing exceptions when a call to refresh_token failed. Rather than doing this, I think we should allow the exception to bubble up to the program using ruby-box (which may or may not choose to attempt refreshing the access token again). To prevent scenarios where we might attempt to refresh a token in an infinite loop, I've added a retries argument to request, only one attempt will be made to update the refresh_token.
- I've made it so when creating a collaboration it defaults the role to 'viewer', you can optionally pass in a different role.

If you have any more updates in the future, could you send them as pull requests from branches -- it was quite a bit of work to pull all these updates in en masse, having said that your contributions are greatly appreciated :thumbsup:

**Pull request merged**